src/shims/unix/fd.rs

```diff
@@ -380,6 +432,42 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         Ok((-1).into())
     }
 
+    fn check_and_update_readiness(&mut self, fd: &mut dyn FileDescription) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        fd.check_readiness(this)
+            .map(|events| self.update_readiness(events, fd.get_epoll_events()?))??;
+        Ok(())
+    }
+
+    /// Function used to update the ready list of an epoll_event. It takes in all the readiness
+    /// flags of a file description and return event as ready if any of the flag is monitored
+    /// by the epoll_event.
+    fn update_readiness(
+        &self,
+        ready_flags: Vec<u32>,
+        epoll_events: &BTreeMap<i32, Weak<EpollEvent>>,
+    ) -> InterpResult<'tcx> {
+        for (_, event) in epoll_events.iter() {
+            // If the current epoll_event contains the flag that is signaled as ready, we add/update
+            // the epoll_return entry in the ready list.
+            if let Some(epoll_event) = event.upgrade() {
+                if let Some(flags) = epoll_event.contains_flag(&ready_flags) {
+                    let weak_file_descriptor = epoll_event.weak_file_descriptor.clone();
+                    let epoll_key = (weak_file_descriptor, epoll_event.file_descriptor);
+                    // Retrieve the epoll return if it is already in the return list.
+                    let ready_list = &mut epoll_event.ready_list.borrow_mut();
+                    // Add a new epoll entry if it doesn't exist, or update the event mask if it exists.
+                    let epoll_return = EpollReturn::new(flags, epoll_event.data);
+                    let epoll_entry = ready_list.entry(epoll_key).or_insert(epoll_return);
+                    // The update here is bitwise or, so it will still be correct if we try to add
+                    // a flag that already exists.
+                    epoll_entry.update_events(flags);
```

Comment:
when `epoll_ctl_mod` changes the `data` field of `EpollEvent`, shouldn't we now also copy the new value into `EpollReturn`?

Reply:
Yup, nice catch!